src/Vault.sol

```diff
@@ -0,0 +1,398 @@
+
+    function _getAssetStorage() internal pure returns (AssetStorage storage $) {
+        assembly {
+            $.slot := 0x2dd192a2474c87efcf5ffda906a4b4f8a678b0e41f9245666251cfed8041e680
+        }
+    }
+
+    function _getStrategyStorage() internal pure returns (StrategyStorage storage $) {
+        assembly {
+            $.slot := 0x36e313fea70c5f83d23dd12fc41865566e392cbac4c21baf7972d39f7af1774d
+        }
+    }
+
+    //// ADMIN ////
+
+    function setRateProvider(address rateProvider_) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (rateProvider_ == address(0)) revert ZeroAddress();
+        _getVaultStorage().rateProvider = rateProvider_;
+        emit SetRateProvider(rateProvider_);
+    }
+
+    function setBufferStrategy(address bufferStrategy_) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (bufferStrategy_ == address(0)) revert ZeroAddress();
+        StrategyStorage storage strategyStorage = _getStrategyStorage();
+        if (!strategyStorage.strategies[bufferStrategy_].active) revert InvalidStrategy();
+        _getVaultStorage().bufferStrategy = bufferStrategy_;
+        emit SetBufferStrategy(bufferStrategy_);
+    }
+
+    function addAsset(address asset_, uint8 decimals_) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (asset_ == address(0)) revert ZeroAddress();
+
+        AssetStorage storage assetStorage = _getAssetStorage();
+        uint256 index = assetStorage.list.length;
+        if (index > 0 && assetStorage.assets[asset_].index != 0) revert InvalidAsset();
+
+        assetStorage.assets[asset_] = AssetParams({active: true, index: index, decimals: decimals_, idleBalance: 0});
+
+        assetStorage.list.push(asset_);
+
+        emit NewAsset(asset_, decimals_, index);
+    }
+
+    function toggleAsset(address asset_, bool active) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        AssetStorage storage assetStorage = _getAssetStorage();
+        if (assetStorage.assets[asset_].index == 0) revert InvalidAsset();
+        assetStorage.assets[asset_].active = active;
+        emit ToggleAsset(asset_, active);
+    }
+
+    function addStrategy(address strategy, uint8 decimals_) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (strategy == address(0)) revert ZeroAddress();
+
+        StrategyStorage storage strategyStorage = _getStrategyStorage();
+        uint256 index = strategyStorage.list.length;
+
+        if (index > 0 && strategyStorage.strategies[strategy].index != 0) {
+            revert DuplicateStrategy();
+        }
+
+        strategyStorage.strategies[strategy] =
+            StrategyParams({active: true, index: index, decimals: decimals_, idleBalance: 0});
+
+        strategyStorage.list.push(strategy);
+
+        emit NewStrategy(strategy, index);
+    }
+
+    function pause(bool paused_) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        VaultStorage storage vaultStorage = _getVaultStorage();
+        if (_getVaultStorage().rateProvider == address(0)) revert RateProviderNotSet();
```

Comment:
the conditionals around pausing are a bit confusing.

it is correct if it starts paused, but in general you'd expect pause/unpause does just that and nothing else

Reply:
the intention was to prevent unpausing unless the rateProvider and bufferStrategy have been setup